xarray/conventions.py

```diff
@@ -389,19 +276,15 @@ def decode_cf_variable(
     if decode_times:
         var = times.CFDatetimeCoder(use_cftime=use_cftime).decode(var, name=name)
 
-    dimensions, data, attributes, encoding = variables.unpack_for_decoding(var)
-    # TODO(shoyer): convert everything below to use coders
+    if decode_endianness and not var.dtype.isnative:
+        var = variables.EndianCoder().decode(var)
+        original_dtype = var.dtype
 
-    if decode_endianness and not data.dtype.isnative:
-        # do this last, so it's only done if we didn't already unmask/scale
-        data = NativeEndiannessArray(data)
-        original_dtype = data.dtype
+    var = variables.BooleanCoder().decode(var)
```

Comment:
Can you add a comment for why the order was changed here? It now comes before `encoding.setdefault("dtype", original_dtype)`

Reply:
I need to check, might be by purpose. I'll come back later with details.

Reply:
`BooleanEncoder ` doesn't need or change contents of `.encoding`. So the order doesn't matter here. In the light of this I've moved this before the unpacking to only have to unpack once before applying `original_dtype` and `LazyIndexedArray`.